user_groups: Support deactivating user groups.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit is a preparatory refactor.
- Second commit is to add `deactivated` field to `NamedUserGroup`.
- Third commit is to add support deactivated groups.
- Fourth commit is to not allow deactivated groups as subgroups, for settings, etc.
- Fifth commit is to allow deactivated groups in API response.
- Last commit is to convert mention of deactivated groups to silent mentions.

 <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on zerver/lib/user_groups.py:
```diff
@@ -114,6 +115,59 @@ def access_user_group_by_id(
+    if NamedUserGroup.objects.filter(
+        realm_id=user_group.realm_id,
+        deactivated=False,
+        can_mention_group_id__in=setting_group_ids_using_deactivating_user_group,
```
But yeah, we would need to add each new setting in the query.

timabbott replied:
I think the single query approach is good; as noted below, we can probably construct a better query by using kwargs to the Django ORM the right way.

sahil839 commented on zerver/lib/user_groups.py:
```diff
@@ -326,6 +346,14 @@ def update_or_create_user_group_for_setting(
+        # whether the group is deactivated or not. Realm ID error would be caught
+        # above and in case the user group is deactivated the error will be raised
+        # in has_user_group_access itself, so there is no coverage here.
+        if not has_user_group_access(subgroup, user_profile, for_read=False, as_subgroup=True):
```
We could have filter out the deactivated groups above in the `NamedUserGroup` query but this seemed better to give a meaningful message mentioning that the group is deactivated.

sahil839 commented on zerver/tests/test_user_groups.py:
```diff
@@ -2037,6 +2121,31 @@ def test_updating_subgroups_of_user_group(self) -> None:
+        # Test that a deactivated group cannot be used as a subgroup.
+        params = {"add": orjson.dumps([support_group.id]).decode()}
+        result = self.client_post(f"/json/user_groups/{test_group.id}/subgroups", info=params)
+        self.assert_json_error(result, "User group is deactivated.")
```
Error message here and in the test for checking user group is used as a setting or not, can be confusing with respect to which group is deactivated. But since we use the `has_user_group_access` function, the error message is this. We can either just add a separate if-condition for deactivation case and have a clear error message explaining which group is deactivated.

timabbott replied:
Probably it's easy to debug in practice. I think it's possible a `code` would be a better way to indicate this than yet another string, if we decide it's important, but this seems OK to procrastinate.

sahil839 commented on zerver/views/user_groups.py:
```diff
@@ -96,8 +96,14 @@ def add_user_group(
+def get_user_group(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    allow_deactivated: bool = REQ(json_validator=check_bool, default=False),
```
Not sure whether this is required or not, but just added for now. Also, do we want to allow deactivated groups in queries like "get members of a group", "get subgroups of a group", etc.

timabbott replied:
Probably a good question for the API design channel; I'm not sure.

timabbott replied:
I think probably this change is definitely good; but you can ask the other ones for a follow-up. 

Also I'm going to rename this `get_user_group` function, I was super confused before realizing its name should be plural.

sahil839 commented:
This is the first part of the groups deactivation part.

Still need to add a commit for a client compatibility to decide whether to pass deactivated groups in `register` response or not. And of course the frontend changes are left.

sahil839 commented on zerver/lib/user_groups.py:
```diff
@@ -119,6 +120,68 @@ def access_user_group_by_id(
+                direct_message_initiator_group_id__in=setting_group_ids_using_deactivating_user_group
+            )
+            | Q(
+                direct_message_permission_group_id__in=setting_group_ids_using_deactivating_user_group
```
Another approach that can be used here is that just fetch group ID for all the settings for the realm in a single query and then checking whether any IDs are common.

timabbott replied:
I suspect that approach might produce cleaner SQL? Not sure, but `get_recursive_subgroups_for_groups` might prove helpful. But my main feedback on this code path is https://github.com/zulip/zulip/pull/30504/commits/447e81bcba3443ef635a9629645320a9165f6203#r1747772019

sahil839 replied:
Updated the code to loop over the settings, and that looks better.

sahil839 commented on zerver/lib/user_groups.py:
```diff
@@ -119,6 +120,68 @@ def access_user_group_by_id(
+        )
+        .exists()
+    ):
+        raise JsonableError(_("You cannot deactivate a user group which is used for setting."))
```
Might be useful to mention the setting name here so that user knows what setting needs to be changed to deactivate the group.

sahil839 replied:
Or, we are also planning to show for what settings a group is being used in the webapp UI, so that might help.

timabbott replied:
Yeah, the setting name would be a nice touch, but it might be complicated to do correctly, and the client should in theory be able to figure that out, so I think it's fine to just have generic errors for this.

timabbott commented on zerver/views/streams.py:
```diff
@@ -575,12 +576,13 @@ def add_subscriptions_backend(
-            setting_name="can_remove_subscribers_group",
-            permission_configuration=permission_configuration,
-        )
+        with transaction.atomic():
```
I'm not sure I understand the strategy for this transaction being so short; it seems like the transaction would end before actually saving the new channel to the database?

sahil839 replied:
This was because `access_user_group_for_setting` uses `select_for_update` and I wanted to avoid other changes which might break something, so just added the transaction for the specific call. But now the whole function is already inside a transaction, so removed that.

timabbott commented on zerver/models/groups.py:
```diff
@@ -60,6 +60,7 @@ class NamedUserGroup(UserGroup):  # type: ignore[django-manager-missing] # djang
     )
 
     realm_for_sharding = models.ForeignKey("zerver.Realm", on_delete=CASCADE, db_column="realm_id")
+    deactivated = models.BooleanField(default=False)
```
Can we use `db_default=False`? I don't see much downside, and it would simplify running the migration to add the field.

sahil839 replied:
Updated.

timabbott commented on zerver/lib/user_groups.py:
```diff
@@ -125,6 +126,75 @@ def access_user_group_by_id(
+    setting_group_ids_using_deactivating_user_group = [
+        *list(anonymous_supergroup_ids),
+        user_group.id,
+    ]
```
This next section probably deserves a comment about how checking whether the group is unused requires checking both whether the group is used directly as the value of a setting or as a subgroup of an anonymous group used for such a setting.

sahil839 replied:
Added the comment.

timabbott commented on zerver/lib/user_groups.py:
```diff
@@ -125,6 +126,75 @@ def access_user_group_by_id(
+        .filter(
+            Q(can_manage_group_id__in=setting_group_ids_using_deactivating_user_group)
+            | Q(can_mention_group_id__in=setting_group_ids_using_deactivating_user_group)
+        )
```
I think we're going to want to write this code to construct the `Q` objects using a loop over `GROUP_PERMISSION_SETTINGS`; that seems like the only mechanism that would be easy to keep correct as we add new settings.... both for this and the below similar blocks for realm/Stream.

sahil839 replied:
Updated.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -3228,6 +3228,12 @@ paths:
+                                    deactivated:
+                                      type: boolean
+                                      description: |
+                                        Whether the user group is deactivated or not.
```
Probably we should say a bit more about what this means. "Whether the user group is deactivated. Deactivated groups cannot be used as a subgroup of another group or used for any other purpose."

sahil839 replied:
Updated.

timabbott commented on zerver/views/user_groups.py:
```diff
@@ -120,7 +120,14 @@ def edit_user_group(
+    if user_group.deactivated and (
+        description is not None or can_mention_group is not None or can_manage_group is not None
+    ):
+        raise JsonableError(_("You can only change name of deactivated user groups"))
```
This check is fragile, in that it won't be correct when adding new settings.

One option would be to, immediately after the `name` block, place this check, so only changing `name` is permitted.

sahil839 replied:
We would need to add the `if user_group.deactivated` check in each `if description is not None`, etc. block so that we do not raise error if only name is changed. Should we do that?

timabbott replied:
I'm not sure. Maybe best to leave this as-is for this PR, and schedule it for a follow-up. But yeah, probably the right model would be to check in each block.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -19869,6 +19869,24 @@ paths:
+                  description: |
+                    Whether to include deactivated user groups in the response.
+
+                    **Changes**: New in Zulip 10.0 (feature level 285).
```
This should just say "Previously, deactivated user groups did not exist and thus would enver be included in the response."?

sahil839 replied:
Updated.

sahil839 commented:
Updated as per feedback. Also, posted follow-up question to one of the comment above.

Need to start #api-design conversation for https://github.com/zulip/zulip/pull/30504#discussion_r1747781471.

timabbott commented:
This looks great; going to go ahead and merge it to keep migration/API changelog numbering under control. I did an edit pass on the API documentation for clarity.

I assume we still need to do some work on the web app to actually use the feature; it seems harmless to have the backend for this merged without the frontend, since at first no groups will be deactivated.

sahil839 commented:
Updated the PR to fix "preferrable" -> "preferable" typo and fixed the feature level update at one place following the rebase.

Keeping auto merge enabled, as there weren't any major changes.

sahil839 commented:
Pushed again to fix the tests, which was a lint failure and was fixed by rebasing the PR to latest main.
